File: schedule_parser.pl
#!/usr/bin/perl

=head1 NAME 

schedule_parser.pl - Create JSON with team info from SIAHL site

=head1 SYNOPSIS

schedule_parser.pl [--html filepath] [--json filepath]

=head1 OPTIONS

=over 8 

=item B<--html>

Use this HTML file instead of pulling from the SIAHL site.

=item B<--json>

Specifies the output path for the JSON created.

=back

=cut

use strict;
use warnings;

use Getopt::Long;
use LWP::UserAgent;
use HTTP::Request;
use JSON;
use Pod::Usage qw(pod2usage);

my %opts;
GetOptions(\%opts,
    'help',
    'html=s',
    'json=s'
) || pod2usage(-verbose => 0);

pod2usage(-verbose => 1) if $opts{help};

my $html_file = $opts{html};
my $html = "";

$opts{json} ||= 'teams.json';

if  ($html_file){
    open(HTML, "< $html_file") || die;
    foreach (<HTML>){ $html .= $_; }
    close(HTML);
}
else{
    my $league_page = "http://stats.liahl.org/display-stats.php?league=1";

    my $ua = new LWP::UserAgent;

    my $req = HTTP::Request->new(GET => $league_page);

    # Pass request to the user agent and get a response back
    my $res = $ua->request($req);

    # Check the outcome of the response
    if ($res->is_success) {
        $html = $res->content;
    }
    else {
        print $res->status_line, "\n";
    }
}

my $teams = get_teams($html);
open(TEAMS, "> $opts{json}") || die "Can't open $opts{json}: $!";
print TEAMS encode_json($teams);
close(TEAMS);

# <TR bgcolor="#00515a"><th colspan=7><a name=93>Senior EEEE</a></th></tr>
# <TR bgcolor="#00515a"><th colspan=7><a href="display-league-stats.php?league=1&level=93&season=14&conf=0">Player Stats</a></th></tr>
# <TR bgcolor="#00515a"><th colspan=7><a href="display-playoff.php?level=93&league=1&season=14&conf=0">Playoff Tree</a></th></tr>
# <tr bgcolor="#00515a"><th>Team</th><th>GP</th><th>W</th><th>L</th><th>T</th><th>OTL</th><th>PTS</th></tr>
# <tr bgcolor="#b4b4b4">
# <td align=center><a href="display-schedule.php?team=298&season=14&tlev=0&tseq=0&league=1">Re-Habs </a></td><td align=center>6</td><td align=center>4</td><td align=center>1</td><td align=center>1</td><td align=center>0</td><td align=center>9</td></tr>
# <tr bgcolor="#e4e4e4">

# ical link for a team
# http://stats.liahl.org/team-cal.php?team=1789&tlev=0&tseq=0&season=17&format=iCal

sub get_teams{
    my @html = split /\n/, shift;
    my $level = "[undefined]";
    my $teams = {};
    foreach (@html){
        if (/name=(\d+)\>\s*Senior\s*([\w ]+)\</){
            $level = $2;     # $1 is the level ID, which we don't need since it's in the team URL already
            $teams->{$level} = {};
        } 
        elsif (/href="display\-schedule.php\?([^"]+)"\>([^<]+)\s*\</){
            my ($url,$name) = ($1,$2);
            $name =~ s/\s+$//;
            $teams->{$level}->{$name} = { url => $url, mtime => '' }; # Stores the URL with all of the team info in it
        }
    }
    return $teams;
}

File: schedule_parser.py
#!/usr/bin/env python

import re
import uuid
import json
import urllib2
from bs4 import BeautifulSoup

# We want to fiter these tags for processing:
#
# Adult level names:
#   <a name="56">Senior A</a>
#
# JSHL level names:
#   <a href="display-schedule.php?league=2&amp;level=12&amp;season=30&amp;conf=2" name="12">Bantam IH Silver</a>
#
# Team pages:
#   <a href="display-schedule.php?team=1463&amp;season=30&amp;tlev=0&amp;tseq=0&amp;league=2">Ranger Blue </a>

def levels_and_teams(tag):
    if tag.name == "a":
        if "name" in tag.attrs:
            return True
        elif "href" in tag.attrs and re.match("display-schedule.php", tag.attrs["href"]):
            return True

def clean_id(bits):
    new_id = ".".join(bits)
    return re.sub(r'[^\w\.]+', '_', new_id)

def new_team_obj(league, level, team, href=None, relcalid=None):
    team_id = clean_id([league, level, team])
    new_team = {
        "mtime": "",
        "filename": "{}.ics".format(team_id),
    }
    new_team["relcalid"] = relcalid if relcalid else str(uuid.uuid5(uuid.NAMESPACE_DNS, str(team_id)))
    if href:
        new_team["url"] = href
    return new_team

class League(object):
    def __init__(self, name, url=None, level_transform=None):
        self.name = name
        self.url  = url
        self.html = None
        self.level_transform = level_transform

    def load(self, html_file=None):
        # TODO: Check for valid file
        infile = open(html_file, "r")
        self.html = infile.read()
        infile.close()

    def fetch(self):
        if self.url is None:
            return
        response = urllib2.urlopen(self.url)
        self.html = response.read()

    def fix_level(self, level_string):
        s = level_string if self.level_transform is None else re.sub(self.level_transform[0], self.level_transform[1], level_string)
        return str(s).rstrip()

    def fix_team(self, team_string):
        return str(team_string).rstrip()

    def fix_team_link(self, href):
        return re.sub(r'^display-schedule\.php\?', '', href)

    def parse(self, relcal_map=None):
        if self.html is None:
            return
        bs = BeautifulSoup(self.html)
        filter_tags = bs.findAll(levels_and_teams)
        levels = {}
        level = None
        for tag in filter_tags:
            # Order here is important since JSHL levels also have hrefs, but we want to match on name
            # {"CC":{"Blueline":{"mtime":"","url":"team=534&season=30&tlev=0&tseq=0&league=1"},
            if tag.text.strip().lower() == self.name.lower():
                # If the league name is a link, skip it
                continue
            if "name" in tag.attrs:
                level = self.fix_level(tag.text)
                if level in levels:
                    raise Exception("Level is defined twice, we got parser troubles, friend")
                levels[level] = {}
            elif "href" in tag.attrs:
                team = self.fix_team(tag.text)
                if level is None:
                    raise Exception("Got a team before a level, something is wrong")
                relcalid = None
                if relcal_map is not None:
                    if level in relcal_map and team in relcal_map[level] and \
                        "relcalid" in relcal_map[level][team]:
                        relcalid = relcal_map[level][team]["relcalid"]
                levels[level][team] = new_team_obj(self.name, level, team, 
                    href=self.fix_team_link(tag.attrs["href"]), relcalid=relcalid)

        return levels


LEAGUES = [
    League("siahl", "http://stats.liahl.org/display-stats.php?league=1", level_transform=[r"^Senior ", ""]),
    League('jshl', "http://stats.liahl.org/display-stats.php?league=2"),
    League("over35", "http://stats.liahl.org/display-schedule.php?league=4"),
]

LEAGUE_LOOKUP = { l.name: l for l in LEAGUES }


if __name__ == '__main__':


    import argparse

    parser = argparse.ArgumentParser(description="Parse SIAHL site for team list")
    parser.add_argument("--html", dest="html", help="HTML to parse")
    parser.add_argument("--league", dest="league", help="League to parse, one of: {}".format(LEAGUE_LOOKUP.keys()))
    parser.add_argument("--out", dest="outfile", help="Ouput filename")
    parser.add_argument("--relcal_json", dest="relcal", help="Existing JSON file to pull relcal IDs from")
    args = parser.parse_args()

    if args.league is None:
        parser.error("--league is required")
    elif args.league not in LEAGUE_LOOKUP:
        parser.error("league '{}' is invalid, should be one of {}".format(args.league, LEAGUE_LOOKUP.keys()))

    league = LEAGUE_LOOKUP[args.league]
    if args.html:
        league.load(args.html)
    else:
        league.fetch()

    parse_args = {}

    # If they gave us a file of existing relcal IDs, parse it and pass it along
    if args.relcal:
        rfile = open(args.relcal, "r")
        relcal_data = rfile.read()
        rfile.close()
        parse_args["relcal_map"] = json.loads(relcal_data)

    parsed_league = league.parse(**parse_args)

    json_txt = json.dumps(parsed_league,sort_keys=True, indent=4, separators=(',', ': '))

    if args.outfile:
        outfile = open(args.outfile, "w")
        outfile.write(json_txt)
        outfile.close()
    else:
        print json_txt
